### PR TITLE
Fixed tab switching too fast when wheeling/scrolling

### LINKED ISSUE #96409: Trackpad: Tab switch on mouse scroll: horizontal scrolling sensitivity too high
Refs: https://github.com/microsoft/vscode/issues/95952

1. "workbench.editor.scrollToSwitchTabs": natural
2. Have a mouse with horizontal press
3. Horizontal scroll is too senstivie in the tab area and seems pretty useless due to that

Vertical scroll works just fine. And horizontal scroll in the editor has good sensitivity so maybe you could check with Alex how he handles the horizontal scroll.

### PR DESCRIPTION
**Pull request by Andre-Fonteles**
This PR fixes #96409 - Trackpad: Tab switch on mouse scroll: horizontal scrolling sensitivity too high. 

PS: I created this new PR to avoid unrelated commits from the old one (#111901).

### CODE REVIEW

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 1]
```diff
@@ -93,6 +93,8 @@ export class TabsTitleControl extends TitleControl {
 
 	private path: IPath = isWindows ? win32 : posix;
 
+	private lastMouseWheelEventTime: number = 0;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:96]:
```suggestion
	private lastMouseWheelEventTime = 0;
```

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 2]
```diff
@@ -337,8 +339,25 @@ export class TabsTitleControl extends TitleControl {
 				}
 			}
 
-			// Figure out scrolling direction
-			const nextEditor = this.group.getEditorByIndex(this.group.getIndexOfEditor(activeEditor) + (e.deltaX < 0 || e.deltaY < 0 /* scrolling up */ ? -1 : 1));
+			// Ignore event if the last one happened too recently.
+			// The restriction is relaxed according to the abs value of deltaX and deltaY
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:343]:
```suggestion
			// The restriction is relaxed according to the absolute value of `deltaX` and `deltaY`
			//  to support discrete (mouse wheel) and contiguous scrolling (touchpad) equally well
```

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 3]
```diff
@@ -337,8 +339,25 @@ export class TabsTitleControl extends TitleControl {
...
+			const time_threshold = 150;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:344]: I suggest to extract this to a `static` constant below [here](https://github.com/microsoft/vscode/blob/f9f2637f794fb728b32d39c25e9eb8a5763afe1d/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L74-L74)

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 4]
```diff
@@ -337,8 +339,25 @@ export class TabsTitleControl extends TitleControl {
...
+
+			// Figure out scrolling direction (ignore direction if the scrolling is too subtile)
+			const distance_threshold = 1.5;
+			let tabSwitchDirection = 0;
+			if (e.deltaX + e.deltaY < - distance_threshold) {
+				tabSwitchDirection = -1;
+			}
+			if (e.deltaX + e.deltaY > distance_threshold) {
+				tabSwitchDirection = 1;
+			}
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:358]:
```suggestion
			if (e.deltaX + e.deltaY < - distance_threshold) {
				tabSwitchDirection = -1;
			} else if (e.deltaX + e.deltaY > distance_threshold) {
				tabSwitchDirection = 1;
			}
```

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 5]
```diff
@@ -337,8 +339,25 @@ export class TabsTitleControl extends TitleControl {
...
+			// Ignore event if the last one happened too recently.
+			// The restriction is relaxed according to the abs value of deltaX and deltaY
+			const time_threshold = 150;
+			if (Date.now() - this.lastMouseWheelEventTime < time_threshold - 2 * (Math.abs(e.deltaX) + Math.abs(e.deltaY))) {
+				return;
+			}
+			this.lastMouseWheelEventTime = Date.now();
+
+			// Figure out scrolling direction (ignore direction if the scrolling is too subtile)
+			const distance_threshold = 1.5;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:351]: I personally prefer camel case names, i.e. `distanceThreshold`

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts [hunk 6]
```diff
@@ -337,8 +339,25 @@ export class TabsTitleControl extends TitleControl {
...
+			let tabSwitchDirection = 0;
+			if (e.deltaX + e.deltaY < - distance_threshold) {
+				tabSwitchDirection = -1;
+			}
+			if (e.deltaX + e.deltaY > distance_threshold) {
+				tabSwitchDirection = 1;
+			}
+			const nextEditor = this.group.getEditorByIndex(this.group.getIndexOfEditor(activeEditor) + tabSwitchDirection);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:359]: Could/should we return early from this method if we don't have a `tabSwitchDirection`? In other words, do we need the code to execute even if the same editor is going to be opened? I would prefer to ignore the work of opening an editor unless it really changes.

**Review by bpasero — CHANGES_REQUESTED** [review]
Tested it with touchpad and mouse and seems to be working great 👍 . I left some feedback and suggested code changes.

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts (hunk 4 above)
Comment (Andre-Fonteles) [reply src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:358]: You're right. For some reason I initially thought both conditions could be true, but that's obviously not the case.

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts (hunk 6 above)
Comment (Andre-Fonteles) [reply src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:359]: You're right, we should return already if `tabSwitchDirection` is zero. I thought the `if` that checks if `nextEditor` is set was taking care of this, but it is not.

File: src/vs/workbench/browser/parts/editor/tabsTitleControl.ts (hunk 5 above)
Comment (Andre-Fonteles) [reply src/vs/workbench/browser/parts/editor/tabsTitleControl.ts]: Me too. I've been using python too much recently... :-)

**Comment by Andre-Fonteles** [comment]
Thanks for the feedback, @bpasero. I adopted all your suggestions.

**Pull request merged**